src/lib/UI/Value/ValueFactory.php

```diff
@@ -158,16 +159,30 @@ public function createLanguage(Language $language, VersionInfo $versionInfo): UI
      *
      * @throws \eZ\Publish\API\Repository\Exceptions\NotFoundException
      * @throws \eZ\Publish\API\Repository\Exceptions\UnauthorizedException
+     * @throws \eZ\Publish\API\Repository\Exceptions\BadStateException
      */
     public function createRelation(Relation $relation, Content $content): UIValue\Content\Relation
     {
         $contentType = $content->getContentType();
         $fieldDefinition = $contentType->getFieldDefinition($relation->sourceFieldDefinitionIdentifier);
 
+        $contentInfo = $content->contentInfo;
+
+        try {
+            $relationLocation = $this->locationService->loadLocation($contentInfo->mainLocationId);
+        } catch (UnauthorizedException $e) {
+            // try different locations if main location is not accessible for the user
+            $relationLocations = $this->locationService->loadLocations($contentInfo);
```

Comment:
It's a bit painful to see to have to do such logic in UI code, maybe we can enhance `->loadLocations($contentInfo);` to:
1. Always return main location first
2. maybe even allow to set limit so we only "hydrate"/map 1  object here